### PR TITLE
[clang] Emit bad shift warnings

### PR DESCRIPTION
**Pull request by budimirarandjelovichtec**
Diagnose bad shifts and emit warnings

### CODE REVIEW

File: clang/lib/Sema/SemaExpr.cpp [hunk 1]
```diff
@@ -11773,7 +11786,25 @@ QualType Sema::CheckShiftOperands(ExprResult &LHS, ExprResult &RHS,
       isScopedEnumerationType(RHSType)) {
     return InvalidOperands(Loc, LHS, RHS);
   }
-  DiagnoseBadShiftValues(*this, LHS, RHS, Loc, Opc, LHSType);
+
+  BadShiftValueKind BSVKind =
+      DiagnoseBadShiftValues(*this, LHS, RHS, Loc, Opc, LHSType);
+  if (ExprEvalContexts.back().isConstantEvaluated()) {
```
Comment (shafik) [inline clang/lib/Sema/SemaExpr.cpp:11792]: This feels pretty hacky but perhaps I am not fully understanding the solution.

Can you update the description with more details on why this is a good approach to the problem.

Comment (budimirarandjelovicsyrmia) [reply clang/lib/Sema/SemaExpr.cpp:11792]: This solution originally was posted on [Phabricator](https://reviews.llvm.org/D141192). I reviewed commit and found that it makes more sense to diagnose bad shift values inside function DiagnoseBadShiftValues(...) instead in mentioned IF. If there are no more questions, I will edit this part of code.

File: clang/lib/AST/ExprConstant.cpp [hunk 2]
```diff
@@ -2803,9 +2803,10 @@ static bool handleIntIntBinOp(EvalInfo &Info, const Expr *E, const APSInt &LHS,
     // C++11 [expr.shift]p1: Shift width must be less than the bit width of
     // the shifted type.
     unsigned SA = (unsigned) RHS.getLimitedValue(LHS.getBitWidth()-1);
-    if (SA != RHS) {
+    if (SA != RHS && Info.getLangOpts().CPlusPlus11) {
```
Comment (AaronBallman) [inline clang/lib/AST/ExprConstant.cpp:2806]: Why are we removing this diagnostic in C++98 and C?

File: clang/lib/AST/ExprConstant.cpp [hunk 3]
```diff
@@ -2836,9 +2837,11 @@ static bool handleIntIntBinOp(EvalInfo &Info, const Expr *E, const APSInt &LHS,
     // C++11 [expr.shift]p1: Shift width must be less than the bit width of the
     // shifted type.
     unsigned SA = (unsigned) RHS.getLimitedValue(LHS.getBitWidth()-1);
-    if (SA != RHS)
+    if (SA != RHS && Info.getLangOpts().CPlusPlus11) {
```
Comment (AaronBallman) [inline clang/lib/AST/ExprConstant.cpp:2840]: Same here?

File: clang/test/C/drs/dr0xx.c [hunk 4]
```diff
@@ -426,7 +426,7 @@ void dr081(void) {
   /* Demonstrate that we don't crash when left shifting a signed value; that's
    * implementation defined behavior.
    */
- _Static_assert(-1 << 1 == -2, "fail"); /* Didn't shift a zero into the "sign bit". */
+ _Static_assert(-1 << 1 == -2, "fail"); /* expected-warning {{shifting a negative signed value is undefined}} */
```
Comment (AaronBallman) [inline clang/test/C/drs/dr0xx.c:429]: Please fix the comment above; it's no longer implementation-defined behavior, it's undefined since C99

File: clang/test/Sema/shift-count-negative.c [hunk 5]
```diff
@@ -0,0 +1,8 @@
+// RUN: %clang_cc1 -x c -fsyntax-only -verify -Wshift-count-negative %s
+// RUN: %clang_cc1 -x c -fsyntax-only -verify -Wall %s
+// RUN: %clang_cc1 -x c++ -fsyntax-only -verify -Wshift-count-negative %s
+// RUN: %clang_cc1 -x c++ -fsyntax-only -verify -Wall %s
+
+enum shiftof {
+    X = (1<<-29) //expected-warning {{shift count is negative}}
```
Comment (AaronBallman) [inline clang/test/Sema/shift-count-negative.c:7]:
```suggestion
    X = (1<<-29) // expected-warning {{shift count is negative}}
```

File: clang/test/Sema/shift-negative-value.c [hunk 6]
```diff
@@ -0,0 +1,8 @@
+// RUN: %clang_cc1 -x c -fsyntax-only -verify -Wshift-negative-value %s
+// RUN: %clang_cc1 -x c -fsyntax-only -verify -Wall %s
+// RUN: %clang_cc1 -x c++ -fsyntax-only -verify -Wshift-negative-value %s
+// RUN: %clang_cc1 -x c++ -fsyntax-only -verify -Wall %s
+
+enum shiftof {
+    X = (-1<<29) //expected-warning {{shifting a negative signed value is undefined}}
```
Comment (AaronBallman) [inline clang/test/Sema/shift-negative-value.c:7]:
```suggestion
    X = (-1<<29) // expected-warning {{shifting a negative signed value is undefined}}
```

**Review by AaronBallman — COMMENTED** [review]
The patch summary doesn't really explain why the changes are needed. Can you explain what problem you're solving (perhaps link to an issue if this is fixing one that was reported by someone)?

**Comment by budimirarandjelovicsyrmia** [comment]
Ping @AaronBallman

**Comment by tbaederr** [comment]
There are two questions above about removing a diagnostic that you haven't answered yet

File: clang/lib/AST/ExprConstant.cpp (hunk 2 above)
Comment (budimirarandjelovicsyrmia) [reply clang/lib/AST/ExprConstant.cpp]: As I didn't find valid reason, I reversed this change.

File: clang/lib/AST/ExprConstant.cpp (hunk 3 above)
Comment (budimirarandjelovicsyrmia) [reply clang/lib/AST/ExprConstant.cpp]: Same

File: clang/lib/AST/ExprConstant.cpp [hunk 7]
```diff
@@ -2805,7 +2805,7 @@ static bool handleIntIntBinOp(EvalInfo &Info, const Expr *E, const APSInt &LHS,
     unsigned SA = (unsigned) RHS.getLimitedValue(LHS.getBitWidth()-1);
     if (SA != RHS) {
       Info.CCEDiag(E, diag::note_constexpr_large_shift)
-        << RHS << E->getType() << LHS.getBitWidth();
+          << RHS << E->getType() << LHS.getBitWidth();
```
Comment (AaronBallman) [inline clang/lib/AST/ExprConstant.cpp:2808]: Only unrelated whitespace changes in the file, so I think the whole file can be reverted.

File: clang/lib/Sema/SemaExpr.cpp [hunk 8]
```diff
@@ -19959,7 +19961,10 @@ bool Sema::DiagRuntimeBehavior(SourceLocation Loc, ArrayRef<const Stmt*> Stmts,
   case ExpressionEvaluationContext::ConstantEvaluated:
   case ExpressionEvaluationContext::ImmediateFunctionContext:
     // Relevant diagnostics should be produced by constant evaluation.
-    break;
+
+    // Solution below works for warnings detected in
+    // Sema::DiagnoseBadShiftValues()
+    return DiagIfReachable(Loc, Stmts, PD);
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaExpr.cpp:19967]: This seems incorrect to me -- `DiagRuntimeBehavior()` is used to diagnose runtime issues, so we should not be emitting diagnostics for constant evaluations.

File: clang/lib/Sema/SemaExpr.cpp [hunk 9]
```diff
@@ -11428,7 +11428,7 @@ static bool isScopedEnumerationType(QualType T) {
   return false;
 }
 
-static void DiagnoseBadShiftValues(Sema& S, ExprResult &LHS, ExprResult &RHS,
+static void DiagnoseBadShiftValues(Sema &S, ExprResult &LHS, ExprResult &RHS,
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaExpr.cpp:11431]:
```suggestion
static void DiagnoseBadShiftValues(Sema& S, ExprResult &LHS, ExprResult &RHS,
```
Backing out an unrelated formatting change.

File: clang/lib/Sema/SemaExpr.cpp [hunk 10]
```diff
@@ -11523,9 +11525,9 @@ static void DiagnoseBadShiftValues(Sema& S, ExprResult &LHS, ExprResult &RHS,
   }
 
   S.Diag(Loc, diag::warn_shift_result_gt_typewidth)
-    << HexResult.str() << Result.getMinSignedBits() << LHSType
-    << Left.getBitWidth() << LHS.get()->getSourceRange()
-    << RHS.get()->getSourceRange();
+      << HexResult.str() << Result.getMinSignedBits() << LHSType
+      << Left.getBitWidth() << LHS.get()->getSourceRange()
+      << RHS.get()->getSourceRange();
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaExpr.cpp:11530]:
```suggestion
    << HexResult.str() << Result.getMinSignedBits() << LHSType
    << Left.getBitWidth() << LHS.get()->getSourceRange()
    << RHS.get()->getSourceRange();
```
Backing out more unrelated formatting changes.

**Review by AaronBallman — CHANGES_REQUESTED** [review]

File: clang/lib/AST/ExprConstant.cpp (hunk 7 above)
Comment (budimirarandjelovicsyrmia) [reply clang/lib/AST/ExprConstant.cpp]: Reverted changes. Just to mention that 'git clang-format' proposed spaces in this comment and comments below.

Comment (AaronBallman) [reply clang/lib/AST/ExprConstant.cpp]: FWIW, you should be using `clang-format-diff` so that you're only formatting the changes made in the PR: https://clang.llvm.org/docs/ClangFormat.html#script-for-patch-reformatting

File: clang/lib/Sema/SemaExpr.cpp [hunk 11]
```diff
@@ -11444,9 +11444,12 @@ static void DiagnoseBadShiftValues(Sema& S, ExprResult &LHS, ExprResult &RHS,
...
   if (Right.isNegative()) {
-    S.DiagRuntimeBehavior(Loc, RHS.get(),
-                          S.PDiag(diag::warn_shift_negative)
-                            << RHS.get()->getSourceRange());
+    if (S.ExprEvalContexts.back().isConstantEvaluated())
+      S.Diag(Loc, diag::warn_shift_negative) << RHS.get()->getSourceRange();
+    else
+      S.DiagRuntimeBehavior(Loc, RHS.get(),
+                            S.PDiag(diag::warn_shift_negative)
+                              << RHS.get()->getSourceRange());
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaExpr.cpp:11452]: I still don't think this is the correct fix -- I think the issue is that ExprConstant.cpp needs to be taught about these as undefined behaviors. Notice how GCC correctly emits the pedantic diagnostic but Clang does not: https://godbolt.org/z/arb8Y9G9q

When I step through in the debugger, it seems that we do catch the issue in `handleIntIntBinOp()` but we're suppressing diagnostics for it in C more broadly. That seems more likely to be the root cause of the issue. (Basically, we want the diagnostics to be split between SemaExpr.cpp for runtime reachable diagnostics and ExprConstant.cpp for constant expressions.)

CC @tbaederr for opinions on constant expression behavior here.

Comment (tbaederr) [reply clang/lib/Sema/SemaExpr.cpp:11452]: The constant interpreter already diagnoses this:
https://github.com/llvm/llvm-project/blob/51089e360e37962c7841fe0a494ba9fb5368bab2/clang/lib/AST/ExprConstant.cpp#L2850-L2851

and for the test case with the `(-1 << 29)` enum value, this is evaluated in `Sema::VerifyIntegerConstantExpresssion()` and it's also diagnosed here:
https://github.com/llvm/llvm-project/blob/51089e360e37962c7841fe0a494ba9fb5368bab2/clang/lib/Sema/SemaExpr.cpp#L18158-L18159

But the diagnostic doesn't show up. Not sure why. If I modify the constant interpreter to `return false` in the negative LHS case, I get the expected diagnostic:
```console
/home/tbaeder/test.cpp:2:9: error: expression is not an integral constant expression
    2 |     X = (-1<<29) // expected-warning {{shifting a negative signed value is undefined}}
      |         ^~~~~~~~
/home/tbaeder/test.cpp:2:12: note: left shift of negative value -1
    2 |     X = (-1<<29) // expected-warning {{shifting a negative signed value is undefined}}
      |            ^
```

Comment (budimirarandjelovicsyrmia) [reply clang/lib/Sema/SemaExpr.cpp]: Diagnose is caught in handleIntIntBinOp() and is emitted for C++ in VerifyIntegerConstantExpression(). There are two reasons why diagnostic could be emitted in case of C++:
	1) evaluating expression as rvalue fails or evaluation result has side effects or it is not an integer; evaluating expression as rvalue fails if function handleIntIntBinOp() returns false
	2) folding is allowed (CanFold == AllowFoldKind::AllowFold); this value depends on place where constant expression is defined; eg. in case on enum folding is allowed and in case of static assertion it is not allowed
Also, diagnostic would not be emitted if Diagnoser is suppresed.

However, in C case there are no diagnostic for integer constant expression.

So I enabled diagnosing in C case (if Diagnoser is not suppressed). In function EvaluateKnownConstIntCheckOverflow() it is expected that evaluating rvalue is successful, so I keep handleIntIntBinOp() returning true.
For C++ diagnostic, I modified handleIntIntBinOp() to return false.

As folding is allowed for enum and not for static assertion, I added in tests to check both enum and static assertion diagnosing.

File: clang/test/C/drs/dr0xx.c [hunk 12]
```diff
@@ -430,7 +430,8 @@ void dr081(void) {
   /* Demonstrate that we don't crash when left shifting a signed value; that's
    * implementation defined behavior.
    */
- _Static_assert(-1 << 1 == -2, "fail"); /* Didn't shift a zero into the "sign bit". */
+ _Static_assert(-1 << 1 == -2, "fail"); /* c89only-error {{static assertion expression is not an integral constant expression}}
```
Comment (AaronBallman) [inline clang/test/C/drs/dr0xx.c:433]: Something seems wrong here -- why is this an error only in C89 mode? This should be pedantically diagnosed as a warning, not an error (in all C language modes).

File: clang/test/C/drs/dr2xx.c [hunk 13]
```diff
@@ -277,7 +277,9 @@ void dr258(void) {
 void dr261(void) {
   /* This is still an integer constant expression despite the overflow. */
   enum e1 {
-    ex1 = __INT_MAX__ + 1  /* expected-warning {{overflow in expression; result is -2'147'483'648 with type 'int'}} */
+    ex1 = __INT_MAX__ + 1  /* expected-warning {{overflow in expression; result is -2'147'483'648 with type 'int'}}
+                              c89only-error {{expression is not an integer constant expression}}
```
Comment (AaronBallman) [inline clang/test/C/drs/dr2xx.c:281]: Same question here as to why this is an error in C89.

File: clang/test/C/drs/dr0xx.c (hunk 12 above)
Comment (budimirarandjelovicsyrmia) [reply clang/test/C/drs/dr0xx.c]: This was error in all C language modes. At first, I added diagnosing this error in C mode (file SemaExpr.cpp, lines between 1732-1750, function diagnoseNotICE). After suggestion, I changed it to pedantically emit warning (substitute diagnoseNotICE with diagnoseFold, same part of code).

**Comment by budimirarandjelovicsyrmia** [comment]
Ping @AaronBallman @shafik @tbaederr

File: clang/test/C/drs/dr2xx.c [hunk 14]
```diff
@@ -277,7 +277,9 @@ void dr258(void) {
 void dr261(void) {
   /* This is still an integer constant expression despite the overflow. */
```
Comment (AaronBallman) [inline clang/test/C/drs/dr2xx.c:278]: Probably good to update this comment to mention "extension".

File: clang/docs/ReleaseNotes.rst [hunk 15]
```diff
@@ -686,6 +686,8 @@ Improvements to Clang's diagnostics
 - Clang no longer emits a "no previous prototype" warning for Win32 entry points under ``-Wmissing-prototypes``.
   Fixes #GH94366.
 
+- Clang now diagnoses non-C++11 integer constant expressions. Fixes #GH59863
```
Comment (AaronBallman) [inline clang/docs/ReleaseNotes.rst:689]:
```suggestion
- Clang now diagnoses integer constant expressions that are folded to a constant value as an extension in more circumstances . Fixes #GH59863
```

**Review by AaronBallman — APPROVED** [review]
LGTM aside from a few small things to correct. @tbaederr are you happy with the state of things as well?

**Review by tbaederr — APPROVED** [review]

**Comment by AaronBallman** [comment]
@budimirarandjelovicsyrmia -- do you need someone to land these changes on your behalf?

**Comment by budimirarandjelovicsyrmia** [comment]
> @budimirarandjelovicsyrmia -- do you need someone to land these changes on your behalf?

I don't understand question. I think that I answered correctly to previous comments.

**Comment by AaronBallman** [comment]
> > @budimirarandjelovicsyrmia -- do you need someone to land these changes on your behalf?
> 
> I don't understand question. I think that I answered correctly to previous comments. Are there any changes that I didn't cover?

Nope, everything looks good as far as the patch goes, I was wondering if you needed me to press the "merge" button or whether you had commit privileges so you could do it yourself.

**Comment by budimirarandjelovicsyrmia** [comment]
> > > @budimirarandjelovicsyrmia -- do you need someone to land these changes on your behalf?
> > 
> > 
> > I don't understand question. I think that I answered correctly to previous comments. Are there any changes that I didn't cover?
> 
> Nope, everything looks good as far as the patch goes, I was wondering if you needed me to press the "merge" button or whether you had commit privileges so you could do it yourself.

Thanks! Someone from my company will merge it.

**Comment by alanzhao1** [comment]
FYI this is causing breakages in chrome - the android NDK has an enum element set equal to `1UL << 32` and we get a compile failure for `--target=i686-linux-android26`. See https://crbug.com/352592408,

code in question is here: https://android.googlesource.com/platform/frameworks/native/+/master/libs/nativewindow/include/android/hardware_buffer.h#329

The android NDK code is wrong, but this no longer compiles (even with `-Wno-error`).

**Comment by AaronBallman** [comment]
> The android NDK code is wrong, but this no longer compiles (even with -Wno-error).

Well that's unfortunate! Thank you for letting us know.

@alanzhao1 do you think it's reasonable for the workaround to only apply to code in system headers, or does the NDK get included as regular headers generally? Do you need us to revert the changes while we resolve this?

@budimirarandjelovicsyrmia, can you please add a workaround so that the NDK code can continue to compile?

**Comment by budimirarandjelovicsyrmia** [comment]
> @budimirarandjelovicsyrmia, can you please add a workaround so that the NDK code can continue to compile?

I will investigate this.

**Comment by alanzhao1** [comment]
> @alanzhao1 do you think it's reasonable for the workaround to only apply to code in system headers, or does the NDK get included as regular headers generally? 

Having it only apply to system headers should be OK - in our case, chrome imports these headers as part of a `--sysroot`, which IIRC treats warnings the same way as `-isystem`

>  Do you need us to revert the changes while we resolve this?

If you can revert while working on a solution, that would be helpful.

> > @budimirarandjelovicsyrmia, can you please add a workaround so that the NDK code can continue to compile?
> 
> I will investigate this.

FYI I noticed that gcc can compile this with `-fpermissive`

**Comment by stbergmann** [comment]
This started to cause
```
$ cat test.c
#include <string.h>
void f(void) {
    char a[strlen("x")];
    (void) a;
}
```
```
$ clang -Wall -fsyntax-only test.c
test.c:3:12: warning: variable length array folded to constant array as an extension [-Wgnu-folding-constant]
    3 |     char a[strlen("x")];
      |            ^~~~~~~~~~~
1 warning generated.
```
to emit a warning, and I'm not sure that's intentional?

**Comment by AaronBallman** [comment]
> This started to cause
> 
> ```
> $ cat test.c
> #include <string.h>
> void f(void) {
>     char a[strlen("x")];
>     (void) a;
> }
> ```
> 
> ```
> $ clang -Wall -fsyntax-only test.c
> test.c:3:12: warning: variable length array folded to constant array as an extension [-Wgnu-folding-constant]
>     3 |     char a[strlen("x")];
>       |            ^~~~~~~~~~~
> 1 warning generated.
> ```
> 
> to emit a warning, and I'm not sure that's intentional?

That is intentional; `strlen` is not a valid constant expression in C, and so that is technically a VLA, except we have an extension to fold it back into a constant.

**Comment by alexfh** [comment]
Just to confirm: is this the intentional outcome of this patch? 
https://gcc.godbolt.org/z/Kf9YK1qM3
```
enum {
    a = 1<<12,
    b = (-1)<<13
};
```
```
<source>:3:9: error: expression is not an integral constant expression
    3 |     b = (-1)<<13
      |         ^~~~~~~~
<source>:3:13: note: left shift of negative value -1
    3 |     b = (-1)<<13
      |             ^
```

Though technically, it's UB before C++20, the diagnostic - "expression is not an integral constant expression" - doesn't sound helpful (how is it not constant or not integral?). And what's worse, is that it's an error rather than a warning that can be disabled.

**Comment by AaronBallman** [comment]
> Just to confirm: is this the intentional outcome of this patch?

Before C++20, left shift of -1 does not produce a mathematical result that's within the range of representable values for the result type, so it's undefined behavior, and therefore not a core constant expression, and therefore invalid as the initializer for an enumeration constant. So I believe the behavior is expected and correct; you can see the UB diagnostic in other contexts: https://gcc.godbolt.org/z/bqj6xhzzr and GCC agrees with Clang in the enumeration case: https://gcc.godbolt.org/z/qjKzeaYTz

> the diagnostic - "expression is not an integral constant expression" - doesn't sound helpful (how is it not constant or not integral?). And what's worse, is that it's an error rather than a warning that can be disabled.

It has integral constants but it's not an integral constant expression as far as the language is concerned. The note tells you why it's not a valid integral constant expression. As for warning vs error; the standard makes the code ill-formed and we treat that as an error generally speaking.

If you can't enable C++20 mode, you can still get the same value in C++17 and earlier with well-formed code: https://gcc.godbolt.org/z/3ra83joqs

**Comment by alexfh** [comment]
I see. Thanks for the explanation and for the suggestion on how to fix that. Actually, the only instance I've seen, had already been fixed upstream (https://hg.openjdk.org/jdk/jdk/rev/71495d579a65), so the impact of this compiler behavior change is quite low in my part of the universe.

**Comment by alexfh** [comment]
Now there's a different and more subtle side-effect of this patch: https://gcc.godbolt.org/z/YP3EfGern

The invalid expression `1 << 59` leads to the compiler silently ignoring the in-class initializer of the static data member.

**Comment by AaronBallman** [comment]
> Now there's a different and more subtle side-effect of this patch: https://gcc.godbolt.org/z/YP3EfGern
> 
> The invalid expression `1 << 59` leads to the compiler silently ignoring the in-class initializer of the static data member.

Huh, I don't think that's an intentional change in this patch, good catch! Adding `-pedantic` to the command line yields `warning: in-class initializer for static data member is not a constant expression; folding it to a constant is a GNU extension` so I think we're missing that constant folding behavior at codegen time. CC @efriedma-quic @rjmccall for opinions

**Comment by efriedma-quic** [comment]
I think the issue is the usage of "Info.EvalStatus.Diag" to determine whether we consider the expression an ICE; that's going to lead to unpredictable results, and it's not how we handle things anywhere else.  Not sure what the goal of that change is.

Separately, Sema::AddInitializerToDecl should probably print diagnostics when it checks for an ICE, instead of just emitting the generic diag::err_in_class_initializer_non_constant/diag::ext_in_class_initializer_non_constant.  Any maybe we should consider enabling ext_in_class_initializer_non_constant by default.

**Comment by alanzhao1** [comment]
> > @alanzhao1 do you think it's reasonable for the workaround to only apply to code in system headers, or does the NDK get included as regular headers generally?
> 
> Having it only apply to system headers should be OK - in our case, chrome imports these headers as part of a `--sysroot`, which IIRC treats warnings the same way as `-isystem`
> 
> > Do you need us to revert the changes while we resolve this?
> 
> If you can revert while working on a solution, that would be helpful.

We (Chrome) no longer need a revert - we patched the NDK locally.

**Comment by efriedma-quic** [comment]
Posted #99579.

**Comment by AaronBallman** [comment]
> > > @alanzhao1 do you think it's reasonable for the workaround to only apply to code in system headers, or does the NDK get included as regular headers generally?
> > 
> > 
> > Having it only apply to system headers should be OK - in our case, chrome imports these headers as part of a `--sysroot`, which IIRC treats warnings the same way as `-isystem`
> > > Do you need us to revert the changes while we resolve this?
> > 
> > 
> > If you can revert while working on a solution, that would be helpful.
> 
> We (Chrome) no longer need a revert - we patched the NDK locally.

Thank you for letting us know!

**Comment by glandium** [comment]
> We (Chrome) no longer need a revert - we patched the NDK locally.

Patching the NDK is not a solution that is nice everywhere, though. And the new warning is an error that can't be disabled.

**Comment by AaronBallman** [comment]
> > We (Chrome) no longer need a revert - we patched the NDK locally.
> 
> Patching the NDK is not a solution that is nice everywhere, though. And the new warning is an error that can't be disabled.

FWIW, we addressed this issue and backported the fix in https://github.com/llvm/llvm-project/pull/100452 -- are you seeing issues with the latest release candidate?

**Comment by efriedma-quic** [comment]
#100452 only fixed the "ignoring the in-class initializer of the static data member" issue.  I don't think we did anything to allow downgrading overflow errors to warnings.

**Comment by glandium** [comment]
> > > We (Chrome) no longer need a revert - we patched the NDK locally.
> > 
> > 
> > Patching the NDK is not a solution that is nice everywhere, though. And the new warning is an error that can't be disabled.
> 
> FWIW, we addressed this issue and backported the fix in #100452 -- are you seeing issues with the latest release candidate?

I don't know about the latest release candidate, but trunk still has the problem.

**Comment by efriedma-quic** [comment]
#102390 "fixes" the Android NDK issue.
